cpp/dynamic_community_detection_module/algorithm/dynamic_community_detection.cpp

```diff
@@ -0,0 +1,370 @@
+  for (const auto node_j_id : in_neighbors_ids) {
+    std::vector<std::uint64_t> labels_j = MostProbableLabels(node_j_id);
+
+    if (IsSubset(labels_i, labels_j)) label_similarity++;
+  }
+
+  int node_i_in_degree = in_neighbors_ids.size();
+
+  return label_similarity <= node_i_in_degree * similarity_threshold;
+}
+
+std::unordered_map<std::uint64_t, double> LabelRankT::Propagate(
+    std::uint64_t node_i_id) {
+  std::unordered_map<std::uint64_t, double> new_label_Ps_i;
+
+  // propagate own probabilities (handle self-loops)
+  for (const auto [label, P] : label_Ps[node_i_id]) {
+    new_label_Ps_i.insert({label, w_selfloop / sum_w[node_i_id] * P});
+  }
+
+  // propagate neighbors’ probabilities
+  for (const auto node_j_id : InNeighborsMemgraphIDs(node_i_id)) {
+    double contribution =
+        GetTotalWeightBetween(node_j_id, node_i_id) / sum_w[node_i_id];
+    for (const auto [label, P] : label_Ps[node_j_id]) {
+      new_label_Ps_i[label] += contribution * P;
+    }
+  }
+
+  return new_label_Ps_i;
+}
+
+void LabelRankT::Inflate(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps, double exponent) {
+  double sum_Ps = 0;
+
+  for (const auto [label, _] : node_label_Ps) {
+    double inflated_node_label_Ps = pow(node_label_Ps[label], exponent);
+    node_label_Ps[label] = inflated_node_label_Ps;
+    sum_Ps += inflated_node_label_Ps;
+  }
+
+  for (const auto [label, _] : node_label_Ps) {
+    node_label_Ps[label] /= sum_Ps;
+  }
+}
+
+void LabelRankT::Cutoff(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps,
+    double min_value) {
+  std::vector<std::uint64_t> to_be_removed;
+
+  for (const auto [label, P] : node_label_Ps) {
+    if (P < min_value) to_be_removed.push_back(label);
+  }
+
+  for (const auto label : to_be_removed) {
+    node_label_Ps.erase(label);
+  }
+}
+
+std::pair<bool, std::uint64_t> LabelRankT::Iteration(
+    bool incremental, std::unordered_set<std::uint64_t> changed_nodes,
+    std::unordered_set<std::uint64_t> to_delete) {
+  std::unordered_map<std::uint64_t, std::unordered_map<std::uint64_t, double>>
+      updated_label_Ps;
+
+  bool none_updated = true;
+  std::uint64_t most_updates = 0;
+
+  for (const auto node_id : NodesMemgraphIDs()) {
+    if (incremental) {
+      bool was_updated = changed_nodes.count(node_id) > 0 ? true : false;
+      bool was_deleted = to_delete.count(node_id) > 0 ? true : false;
+
+      if (!was_updated || was_deleted) continue;
+    }
+
+    // node selection (a.k.a. conditional update)
+    if (!DistinctEnough(node_id, similarity_threshold)) continue;
+    none_updated = false;
+
+    // label propagation
+    std::unordered_map<std::uint64_t, double> updated_node_label_Ps =
+        Propagate(node_id);
+
+    // inflation
+    Inflate(updated_node_label_Ps, exponent);
+
+    // cutoff
+    Cutoff(updated_node_label_Ps, min_value);
+
+    updated_label_Ps.insert({node_id, updated_node_label_Ps});
```

Comment:
```suggestion
    updated_label_Ps.insert({node_id, std::move(updated_node_label_Ps)});
```

Reply:
What is the rationale for this change? Just wondering if there’s something I’m not seeing here

Reply:
Yes! by moving it you can spare a copy! The new item that is inserted into `updated_label_Ps` will "steal" the resources of `updated_node_label_Ps`. I will write a explanation about what move is shortly.